### PR TITLE
8343177: JFR: Remove critical section for thread id assignment

### PR DESCRIPTION
**Pull request by mgronlun**
Greetings,

With Loom, JFR had to change the assignment of unique thread IDs to threads. For JavaThreads, a dependency exists on the threadObj before the thread ID can be determined and assigned. We currently have a lazy assignment scheme that assigns on first use. Several threads, such as thread iterators and sampling threads, can issue the first use. Hence, a critical section protects assignments.

However, a critical section at this location makes it challenging to build robust signal handlers, for example, because we cannot read the thread ID.

We can remove this critical section with careful rearrangements, ensuring threads are only assigned a thread ID in thread state _thread_new (a thread state that at least the JFR sampler and the JFR iterators exclude).

The problem child is JNI attaching threads, created directly into state _thread_in_vm and added to the threads list before the creation and assignment of the threadObj. We can also manage this case by being careful when we sample such a thread, alternatively allowing for a thread ID of 0 or taking account of the 'attaching via jni' property. 

Testing: jdk_jfr Tier 1- 3, Stress testing

Thanks
Markus

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343177](https://bugs.openjdk.org/browse/JDK-8343177): JFR: Remove critical section for thread id assignment (**Enhancement** - P3)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21756/head:pull/21756` \
`$ git checkout pull/21756`

Update a local copy of the PR: \
`$ git checkout pull/21756` \
`$ git pull https://git.openjdk.org/jdk.git pull/21756/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21756`

View PR using the GUI difftool: \
`$ git pr show -t 21756`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21756.diff">https://git.openjdk.org/jdk/pull/21756.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21756#issuecomment-2443841338)
</details>

### CODE REVIEW

File: src/hotspot/share/runtime/thread.hpp [hunk 1]
```diff
@@ -500,9 +500,9 @@ class Thread: public ThreadShadow {
     return is_in_stack_range_incl(adr, os::current_stack_pointer());
   }
 
-  // Sets this thread as starting thread. Returns failure if thread
+  // Sets the argument thread as starting thread. Returns failure if thread
   // creation fails due to lack of memory, too many threads etc.
-  bool set_as_starting_thread();
+  static bool set_as_starting_thread(JavaThread* jt);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:505]: Why change this? Overall I don't understand the "starting thread" changes.

File: src/hotspot/share/jfr/support/jfrThreadLocal.cpp [hunk 2]
```diff
@@ -421,19 +427,31 @@ traceid JfrThreadLocal::external_thread_id(const Thread* t) {
...
-inline traceid load_java_thread_id(const Thread* t) {
+static inline traceid load_java_thread_id(const Thread* t) {
   assert(t != nullptr, "invariant");
   assert(t->is_Java_thread(), "invariant");
   oop threadObj = JavaThread::cast(t)->threadObj();
   return threadObj != nullptr ? AccessThreadTraceId::id(threadObj) : 0;
 }
 
+#ifdef ASSERT
+static bool assignment_precondition(const Thread* t, JfrThreadLocal* tl) {
```
Comment (dholmes-ora) [inline src/hotspot/share/jfr/support/jfrThreadLocal.cpp:438]:
```suggestion
static bool can_assign(const Thread* t, JfrThreadLocal* tl) {
```

File: src/hotspot/share/jfr/support/jfrThreadLocal.cpp [hunk 3]
```diff
@@ -100,6 +99,13 @@ const JfrBlobHandle& JfrThreadLocal::thread_blob() const {
   return _thread;
 }
 
+void JfrThreadLocal::initialize_primordial_thread(JavaThread* jt) {
+  assert(jt != nullptr, "invariant");
+  assert(Thread::is_starting_thread(jt), "invariant");
+  assert(jt->threadObj() == nullptr, "invariant");
+  jt->jfr_thread_local()->_jvm_thread_id = 1;
```
Comment (dholmes-ora) [inline src/hotspot/share/jfr/support/jfrThreadLocal.cpp:106]: Is this supposed to reflect the `java.lang.Thread` thread-id? If so this is changing to 3 with the object monitor virtual thread changes.

**Review by dholmes-ora — COMMENTED** [review]
Sorry I struggled to map the code changes to the description of the changes. I think I'm missing some background context here - what is the "traceid" of a thread?

**Comment by mgronlun** [comment]
> Sorry I struggled to map the code changes to the description of the changes. I think I'm missing some background context here - what is the "traceid" of a thread?

A traceid is a unique, monotonic ID that JFR assigns to artifacts (instances) (Threads, Klasses, Methods, Symbols).

Invariant; no reuse of IDs.

With Loom, unique monotonic ID is intrinsic to the tid field of the threadObj.

File: src/hotspot/share/jfr/support/jfrThreadLocal.cpp (hunk 3 above)
Comment (mgronlun) [reply src/hotspot/share/jfr/support/jfrThreadLocal.cpp:106]: If the primordial thread ID assignment is changed from 1 to 3, then that will have to be updated, of course.

File: src/hotspot/share/jfr/support/jfrThreadLocal.cpp (hunk 2 above)
Comment (mgronlun) [reply src/hotspot/share/jfr/support/jfrThreadLocal.cpp:438]: Good suggestion.

File: src/hotspot/share/runtime/thread.hpp (hunk 1 above)
Comment (mgronlun) [reply src/hotspot/share/runtime/thread.hpp:505]: I changed this with the argument that it seemed very unbalanced to have a non-static instance method available on all JavaThreads for calling t->set_as_starting_thread(), for setting a static field, a static field which further is only set once, by the main thread during startup in Threads::create_VM().

The assert method is_starting_thread(Thread*t) is for asserting the special assignment of 1 to the primordial thread.

Comment (dholmes-ora) [reply src/hotspot/share/runtime/thread.hpp:505]: Okay so nothing inherently related to the critical section removal, just an API cleanup.

Comment (dholmes-ora) [reply src/hotspot/share/runtime/thread.hpp:505]: Actually the more I look at this the more I dislike the fact `set_as_starting_thread` also calls `os::create_main_thread`. Seems to me that should be done directly by the caller, and if successful we call `set_as_starting_thread`. But that is a different cleanup.

File: src/hotspot/share/jfr/jfr.cpp [hunk 4]
```diff
@@ -100,6 +100,10 @@ void Jfr::on_set_current_thread(JavaThread* jt, oop thread) {
   JfrThreadLocal::on_set_current_thread(jt, thread);
 }
 
+void Jfr::initialize_primordial_thread(JavaThread* jt) {
+  JfrThreadLocal::initialize_primordial_thread(jt);
```
Comment (dholmes-ora) [inline src/hotspot/share/jfr/jfr.cpp:104]: The main thread is generally not the process primordial thread. Can we use main_thread or starting_thread here to keep semi-consistent with the Thread/Threads code terminology? Thanks

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Overall looks good. Thanks for the clarifications.

I have one change request with regards to naming below.

Thanks

File: src/hotspot/share/runtime/thread.hpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/thread.hpp:505]: Just FTR `set_as_starting_thread` originally only called `os::create_main_thread` so why it even existed is a bit of a mystery. The actual `_starting_thread` variable was something I added under JDK-8214097.

File: src/hotspot/share/jfr/support/jfrThreadLocal.cpp (hunk 3 above)
Comment (mgronlun) [reply src/hotspot/share/jfr/support/jfrThreadLocal.cpp:107]: What happens to ID 1 and 2? Other threads that need to be tracked, especially in a similar manner that is done here?

File: src/hotspot/share/jfr/jfr.cpp (hunk 4 above)
Comment (mgronlun) [reply src/hotspot/share/jfr/jfr.cpp:104]: Ok.

Comment (mgronlun) [reply src/hotspot/share/jfr/jfr.cpp]: Updated.

File: src/hotspot/share/jfr/support/jfrThreadLocal.cpp (hunk 3 above)
Comment (mgronlun) [reply src/hotspot/share/jfr/support/jfrThreadLocal.cpp:107]: OK, they are reserved constants that do not denote a thread but particular states.

That's okay from our perspective.

**Review by dholmes-ora — APPROVED** [review]

**Comment by mgronlun** [comment]
Thanks @dholmes-ora for your review.

**Comment by dholmes-ora** [comment]
@mgronlun  - two reviews required for hotspot changes.

**Pull request closed without merging**